machinelearning/tensorflow.md

```diff
@@ -0,0 +1,405 @@
+```
+aws ec2 run-instances --region us-west-2 --image-id <ami-id from the above command>  --instance-type c7g.4xlarge --count 1  --key-name <key name> --subnet-id <subnet_id> --block-device-mappings 'DeviceName=/dev/sda1,Ebs={VolumeSize=256}'
+```
+Once Graviton instance is launched with the above DLAMI, the platform is ready for building Tensorflow based ML inference applications.The DLAMI comes with the additional libraries preinstalled, eg: rust compilers, transformers etc, to enable transformers based (eg: bert) inferencing applications.
+
+
+# Runtime configurations for optimal performance
+Once the Tensorflow setup is ready, enable the below runtime configurations to achieve the best performance.
+
+```
+# The default runtime backend for tensoflow is eigen, but typically onednn+acl provides better performance and this can be enabled by setting the below TF environment variable
+export TF_ENABLE_OENDNN_OPTS=1
+
+# Graviton3 (eg: c7g instances) supports BF16 format for ML acceleration. This can be enabled in oneDNN by setting the below environment variable
+export DNNL_DEFAULT_FPMATH_MODE=BF16
+
+# Tensorflow's inter and intra_op_parallelism_threads and the system openmp thread settings are critical for the optimal workload parallelization in a multi-threaded system.
+# set the inter and intra op thread count during the session creation
+session = Session(
+                 config=ConfigProto(
+                      intra_op_parallelism_threads=<num. of vcpus>,
+                      inter_op_parallelism_threads=1,
+                )
+)
+
+# Make sure the openmp threads are distributed across all the processes for multi process applications to avoid over subscription for the vcpus.
+export OMP_NUM_THREADS = <no. of vcpus> / <no. of processes>
+export OMP_PROC_BIND=false
+export OMP_PLACES=cores
+
+```
+Tensorflow recommends the graph optimization pass for inference to remove training specific nodes, fold batchnorms and fuse operators. This is a generic optimizaion across CPU, GPU or TPU inference, and the optimization script is part of the Tensorflow python [tools](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference_lib.py). For a detailed description, please refer to the Tensorflow [Grappler](https://www.tensorflow.org/guide/graph_optimization) documentation. Below is a snippet of what libraries to import and how to invoke the Grappler passes for inference.
+
+```
+from tensorflow.python.tools.optimize_for_inference_lib import optimize_for_inference
+
+graph_def = tf.compat.v1.GraphDef()
+with tf.compat.v1.gfile.FastGFile(model_path, "rb") as f:
+     graph_def.ParseFromString(f.read())
+
+optimized_graph_def = optimize_for_inference(graph_def, [item.split(':')[0] for item in inputs],
+                    [item.split(':')[0] for item in outputs], dtypes.float32.as_datatype_enum, False)
+g = tf.compat.v1.import_graph_def(optimized_graph_def, name='')
+```
+
```

Comment:
Can you forward reference XLA setup here as the method to get max performance?

Reply:
sure.